hazelcast-jet-core/src/main/java/com/hazelcast/jet/windowing/SlidingWindowP.java

```diff
@@ -100,34 +102,38 @@ protected boolean tryProcessPunc0(@Nonnull Punctuation punc) {
     }
 
     private Traverser<Object> slidingWindowTraverser(Punctuation punc) {
-        Traverser<Object> traverser = Traversers.traverseStream(
-                range(nextFrameSeqToEmit, nextFrameSeqToEmit = wDef.higherFrameSeq(punc.seq()), wDef.frameLength())
-                        .mapToObj(frameSeq ->
-                                computeWindow(frameSeq, seqToKeyToFrame.remove(frameSeq - wDef.windowLength()))
-                                        .entrySet().stream()
-                                        .map(e -> new Frame<>(frameSeq, e.getKey(), finishF.apply(e.getValue()))))
-                        .flatMap(identity()));
-
-        traverser = traverser.append(punc);
-        return traverser;
+        long rangeStart = nextFrameSeqToEmit;
+        nextFrameSeqToEmit = wDef.higherFrameSeq(punc.seq());
+        return Traversers.traverseStream(range(rangeStart, nextFrameSeqToEmit, wDef.frameLength()).boxed())
+                .flatMap(frameSeq -> Traversers.traverseIterable(computeWindow(frameSeq).entrySet())
+                        .map(e -> (Object) new Frame<>(frameSeq, e.getKey(), finishF.apply(e.getValue())))
+                        .onFirstNull(() -> completeWindow(frameSeq)))
+                .append(punc);
     }
 
-    private Map<K, F> computeWindow(long frameSeq, Map<K, F> evictedFrame) {
+    private void completeWindow(long frameSeq) {
+        Map<K, F> evictedFrame = seqToKeyToFrame.remove(frameSeq - wDef.windowLength() + wDef.frameLength());
         if (deductF != null) {
-            // add leading-edge frame
```

Comment:
I find your claim of "doubling the memory usage" unjustified. The smallest number of frames per sliding window is 2. In this case, master code peaks at 3 frames per key, after this PR it's 2 frames. That's a 33% saving, but it assumes no disorder in events. With disorder, there will be more like 4-5 frames with master code, 3-4 frames in case of this PR. So the difference is now 25% or less, and we are still discussing the worst case which is also quite unrealistic (the window is much more likely to slide by a smaller fraction of its length).